### PR TITLE
provider/aws: Add support for api_gateway_method_settings

### LINKED ISSUE #6612: aws_api_gateway_deployment doesn't support stage configuration
Terraform v0.6.16

Beyond stage variables, `aws_api_gateway_deployment` doesn't provide a way to configure other settings available in the console, such as "Cache Settings", "CloudWatch Settings", and "Default Method Throttling".  In order to enable logging, I have to manually make changes in the console.

### LINKED ISSUE #6972: Add aws_api_gateway_stage to allow enabling CloudWatch metrics per method
### Terraform Version

0.6.16
### Affected Resource(s)

aws_api_gateway_method
### Feature Request

I would like to be able to specify whether or not CloudWatch metrics are enabled for a given API Gateway method. This is currently possible through the AWS UI and API, as well as through the Go SDK (http://docs.aws.amazon.com/sdk-for-go/api/service/apigateway.html#type-UpdateMethodInput).
### Notes

It does not appear to me to be possible to set whether metrics are enabled while creating the method (other than through the UI). It seems necessary to make an update call after creation time.
### Use case

I have an API being powered by API Gateway. I want the metrics associated with the API Gateway components to be useable in my third party metrics dashboard. Unfortunately, I cannot export them to Graphite until they exist in CloudWatch. Currently, I am working around this by running a Python script to enable the CloudWatch metrics after all of my `terraform apply` invocations.

### PR DESCRIPTION
**Pull request by radeksimko**
Closes #6972
Closes #6612

### Test plan

```
make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSAPIGatewayMethodSettings_'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/04/11 17:21:59 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSAPIGatewayMethodSettings_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayMethodSettings_basic
--- PASS: TestAccAWSAPIGatewayMethodSettings_basic (108.34s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	108.374s
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/apigateway"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsApiGatewayMethodSettings() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsApiGatewayMethodSettingsUpdate,
```
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: There's no difference between `Create` and `Update`?

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+	methodPath := d.Get("method_path").(string)
+	settings, ok := stage.MethodSettings[methodPath]
+	if !ok {
+		log.Printf("[WARN] API Gateway Method Settings for %q not found, removing", methodPath)
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("settings.0.metrics_enabled", settings.MetricsEnabled)
```
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: Do you think they'll/we'll ever allow more than one settings object at a time? If so, might be worth it to go ahead and write a flatten function here. If not, ignore.

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+	methodPath := d.Get("method_path").(string)
+	prefix := fmt.Sprintf("/%s/", methodPath)
+
+	ops := make([]*apigateway.PatchOperation, 0)
+	if d.HasChange("settings.0.metrics_enabled") {
+		ops = append(ops, &apigateway.PatchOperation{
+			Op:    aws.String("replace"),
+			Path:  aws.String(prefix + "metrics/enabled"),
+			Value: aws.String(fmt.Sprintf("%t", d.Get("settings.0.metrics_enabled").(bool))),
```
Comment (grubernaut) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: Just double checking that `Value` needs to be a `string` here?

**Comment by apparentlymart** [comment]
This sort of resource that updates something that already exists rather than creating a new thing can be confusing. I assume there's a reason not to just add these settings into the `api_gateway_method` resource itself?

File: website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Provides an API Gateway Method Settings, e.g. logging or monitoring.
+
+## Example Usage
+
+```hcl
+resource "aws_api_gateway_method_settings" "s" {
+  rest_api_id = "${aws_api_gateway_rest_api.test.id}"
+  stage_name  = "${aws_api_gateway_stage.test.stage_name}"
+  method_path = "${aws_api_gateway_resource.test.path_part}/${aws_api_gateway_method.test.http_method}"
```
Comment (apparentlymart) [inline website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown]: Since it seems like this is just being used to construct a prefix for a longer path anyway (and thus `method_path` is not a first-class concept in the underlying API?), perhaps it would be better to leave these things separated and combine them in code, thus avoiding the need to explain it below as just "concatenate two things together".

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go (hunk 1 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: There is _no_ create method, because the method settings is part of the stage which we're updating. The `aws_api_gateway_stage` resource in turn has no notion of method settings so they shouldn't clash in any way.

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go (hunk 2 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: I doubt it given the nature of these settings (e.g. on/off logging or monitoring) and even if they do so I think we should keep it as is because the ability to use `count` on top level resources allows the user to do all kinds of things they wouldn't be able to do if they worked with "subresources" (multiple `settings` blocks nested inside the resource).

File: website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown (hunk 4 above)
Comment (radeksimko) [inline website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown]: Thanks for questioning this - I was actually playing with that idea too! 😄 

I'm not completely opposed to 2 fields (e.g. `path_part` & `method`).

The API allows you to define things like `*/*` or `*/GET` and I'm just little bit worried that they might later introduce (or even already support) some kind of special magic syntax which doesn't involve `/` and we'd prevent the user from specifying such paths. 🤔

File: builtin/providers/aws/resource_aws_api_gateway_method_settings.go (hunk 3 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_api_gateway_method_settings.go]: It surprised me too when I saw that for the first time, but I guess it makes sense if they want to support patching of whole objects (marshalled JSON which is just string), not just primitive data types.

TL;DR: https://github.com/aws/aws-sdk-go/blob/master/service/apigateway/api.go#L17073-L17077

File: website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown (hunk 4 above)
Comment (Porkbutts) [reply website/source/docs/providers/aws/r/api_gateway_method_settings.html.markdown]: **path_part** won't work if the resource has a non-root parent though, right? In that case I ended up using **path** which contains the fully qualified path, but the leading slash returned in the path was causing the settings to not be applied properly. My solution for the time being looks like:

```
"${substr("${aws_api_gateway_resource.test.path}/${aws_api_gateway_method.test.http_method}", 1, -1)}"
```
